neo/public/javascripts/common/services.js

```diff
@@ -32,50 +57,183 @@ angular.module('cloudberry.common', [])
+      },
+
+      byHashTagRequest: function(parameters) {
+        return {
+          dataset: parameters.dataset,
+          filter: this.getFilter(parameters, defaultNonSamplingDayRange),
+          unnest: [{
+            hashtags: "tag"
+          }],
+          group: {
+            by: [{
+              field: "tag"
+            }],
+            aggregate: [{
+              field: "*",
+              apply: {
+                name: "count"
+              },
+              as: "count"
+            }]
           },
-          timeBin : parameters.timeBin,
-          geoLevel: parameters.geoLevel,
-          geoIds : parameters.geoIds
-        }));
-        ws.send(json);
+          select: {
+            order: ["-count"],
+            limit: 50,
+            offset: 0
+          }
+        };
+      },
+
+      getFilter: function(parameters, maxDay) {
+        var spatialField = this.getLevel(parameters.geoLevel);
+        var keywords = [];
+        for(var i = 0; i < parameters.keywords.length; i++){
+          keywords.push(parameters.keywords[i].replace("\"", "").trim());
+        }
+        var queryStartDate = new Date(parameters.timeInterval.end);
+        queryStartDate.setDate(queryStartDate.getDate() - maxDay);
+        queryStartDate = parameters.timeInterval.start > queryStartDate ? parameters.timeInterval.start : queryStartDate;
+
+        return [
+            {
+              field: "geo_tag." + spatialField,
+              relation: "in",
+              values: parameters.geoIds
+            }, {
+              field: "create_at",
+              relation: "inRange",
+              values: [queryStartDate.toISOString(), parameters.timeInterval.end.toISOString()]
+            }, {
+              field: "text",
+              relation: "contains",
+              values: [this.mkString(keywords, ",")]
+            }
+        ];
+      },
+
+      getLevel: function(level){
+        switch(level){
+          case "state" : return "stateID";
+          case "county" : return "countyID";
+          case "city" : return "cityID";
+        }
+      },
+
+      mkString: function(array, delimiter){
+        var s = "";
+        array.forEach(function (item) {
+            s += item.toString() + delimiter;
+        });
+        return s.substring(0, s.length-1);
       }
     };
 
     ws.onmessage = function(event) {
       $timeout(function() {
         var result = JSONbig.parse(event.data);
-        switch (result.key) {
-          case "byPlace":
-            asterixService.mapResult = result.value;
-            break;
-          case "byTime":
-            asterixService.timeResult = result.value;
-            break;
-          case "byHashTag":
-            asterixService.hashTagResult = result.value;
-            break;
+        switch (result.transform.wrap.key) {
           case "sample":
-            asterixService.tweetResult = result.value[0];
+            asterixService.tweetResult = result.transform.wrap.value[0];
```

Comment:
the result should not inside the `transform`, the result should be just one record
```
   {
       key : `sampel`, // or else
       value : [ ]    // a list of result
   }
```